src/IO/readers.jl

```diff
@@ -0,0 +1,356 @@
+end
+
+function HDF5Reader(filename::AbstractString)
+    file = h5open(filename, "r")
+    cache = Dict()
+    return HDF5Reader(file, cache)
+end
+
+"""
+    Base.read!(reader, path)
+
+Checks the cache to verify if the object, specified by the `path`, is in the cache.
+If it has not been read, this function reads it to the cache and returns the object.
+"""
+function Base.read!(reader::HDF5Reader, path::AbstractString)
+    (; file, cache) = reader
+    name = split(path, "/")[end]
+    if !haskey(cache, name)
+        @assert haskey(file, path)
+        read_new!(reader, name, _type(reader, path))
+    end
+    return cache[name]
+end
+
+"""
+    _type(reader, path)
+
+Extracts the type of the object, specified by the `path`.
+"""
+_type(reader::HDF5Reader, path::AbstractString) =
+    eval(Meta.parse(attrs(reader.file[path])["type"]))
+
+"""
+    matrix_to_cartesianindices(elemorder_matrix)
+
+Converts the `elemorder_matrix` to cartesian indices.
+"""
+function matrix_to_cartesianindices(elemorder_matrix)
+    m, ndims = size(elemorder_matrix)
+    dims = [maximum(elemorder_matrix[:, dim]) for dim in 1:ndims]
+    elemorder = Vector{CartesianIndex{ndims}}(undef, m)
+    for i in 1:m
+        elemorder[i] = CartesianIndex(elemorder_matrix[i, :]...)
+    end
+    return reshape(elemorder, dims...)
+end
+
+"""
+    read_new!(reader, name, ::Type{IntervalDomain})
+
+Reads an object named 'name' of type 'IntervalDomain'.
+"""
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{IntervalDomain},
+)
+    g = reader.file["domains/" * name]
+    coord_type = replace(attrs(g)["coord_type"], "ClimaCore." => "")
+    boundary_names =
+        haskey(attributes(g), "boundary_names") ?
+        eval(Meta.parse(attrs(g)["boundary_names"])) : nothing
+    periodic = boundary_names === nothing ? true : false
+    CoordType = eval(Meta.parse(coord_type))
+    reader.cache[name] = Domains.IntervalDomain(
+        CoordType(eval(Meta.parse(attrs(g)["coord_min"]))),
+        CoordType(eval(Meta.parse(attrs(g)["coord_max"]))),
+        periodic = periodic,
+        boundary_names = boundary_names,
+    )
+    return nothing
+end
+
+"""
+    read_new!(reader, name, ::Type{SphereDomain})
+
+Reads an object named 'name' of type 'SphereDomain'.
+"""
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{SphereDomain},
+)
+    g = reader.file["domains/" * name]
+    reader.cache[name] = Domains.SphereDomain(attrs(g)["radius"])
+    return nothing
+end
+
+"""
+    read_new!(reader, name, ::Type{IntervalMesh})
+
+Reads an object named 'name' of type 'IntervalMesh'.
+"""
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{IntervalMesh},
+)
+    (; file, cache) = reader
+    g = file["meshes/" * name]
+    FT = attrs(g)["FT"] # Float type
+    faces_string = replace(attrs(g)["faces"], "ZPoint" => "ZPoint{$FT}")
+    faces = eval(Meta.parse(faces_string))
```

Comment:
Again, if we write it as an `Array` we shouldn't need to `eval` here

Reply:
A test is added for stretched mesh. However, the last element of the stretched mesh seems to be a NaN, causing problems with the test.

Reply:
Bug fix added, resolved.